### PR TITLE
[schooltv] Add extractor for SchoolTV playlists

### LINKED ISSUE #8163: Site request: schooltv.nl
It belongs to the npo.nl network of sites, so I suspect you can use the same code.
[Example url](http://www.schooltv.nl/video/ademhaling-de-hele-dag-haal-je-adem-maar-wat-gebeurt-er-dan-eigenlijk-in-je-lichaam/)
The ID of the video can be found in the `data-mid` attribute of a div with class `VideoPlayer`.

### PR DESCRIPTION
**Pull request by rrooij**
This closes #8163

### CODE REVIEW

File: youtube_dl/extractor/npo.py [hunk 1]
```diff
@@ -406,6 +406,42 @@ def _real_extract(self, url):
         }
 
 
+class SchoolTVIE(NPOIE):
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: No need to inherit from `NPOIE`.

File: youtube_dl/extractor/npo.py [hunk 2]
```diff
@@ -406,6 +406,42 @@ def _real_extract(self, url):
...
+    IE_NAME = 'schooltv'
+    _VALID_URL = r'https?://(?:www\.)?schooltv\.nl/video/(?P<id>.*[^/])?/'
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: `id` should not be optional. No need in trailing `/`.

File: youtube_dl/extractor/npo.py [hunk 3]
```diff
@@ -406,6 +406,42 @@ def _real_extract(self, url):
...
+        }
+    ]
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+
+        entries = [
+            self.url_result('npo:%s' % video_id)
+            for video_id in re.findall(r'data-mid="([^"]+)"', webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: This is too broad and detects the same video twice.

File: youtube_dl/extractor/npo.py [hunk 4]
```diff
@@ -406,6 +406,42 @@ def _real_extract(self, url):
...
+            }
+        },
+        {
+            'url': 'http://www.schooltv.nl/video/waarom-zijn-beroemde-kunstenaars-vaak-mannen/',
+            'playlist_count': 2,
+            'info_dict': {
+                'id': 'waarom-zijn-beroemde-kunstenaars-vaak-mannen',
+                'title': 'Schooltv: ARTclips - Waarom zijn beroemde kunstenaars vaak mannen?',
+            }
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: If both tests test the same scenario leave only one.

**Comment by rrooij** [comment]
Wow, thanks for the feedback! Should I fix the issues and squash it in the commit?

**Comment by dstftw** [comment]
Yes.

**Comment by rrooij** [comment]
Done.

One question: is it normal behavior that the video file in the test returns a m3u8 playlist? The VPRO IE returns a .m4v and they are both from NPO, so I'm curious whether this is correct.

File: youtube_dl/extractor/npo.py [hunk 5]
```diff
@@ -406,6 +406,32 @@ def _real_extract(self, url):
         }
 
 
+class SchoolTVIE(InfoExtractor):
+    IE_NAME = 'schooltv'
+    _VALID_URL = r'https?://(?:www\.)?schooltv\.nl/video/(?P<id>.*[^/])/?'
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: Change to `(?P<id>[^/]+)`.

File: youtube_dl/extractor/npo.py [hunk 6]
```diff
@@ -406,6 +406,32 @@ def _real_extract(self, url):
...
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+
+        entries = [
+            self.url_result('npo:%s' % video_id)
+            for video_id in re.findall(r'VideoPlayer.*data-mid="([^"]+)"', webpage)
+        ]
+        playlist_title = self._og_search_title(webpage)
+        return self.playlist_result(entries, playlist_id, playlist_title)
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: From what I've seen there is always only one video on the page thus no need in playlist.

**Comment by dstftw** [comment]
>test returns a m3u8

Fixed in a7685b3a6b02be62ebd53e2978ab43b1cde3ec4b.

**Comment by rrooij** [comment]
I think it is fixed by now.

* It's much shorter now that it doesn't try to find a playlist (it is pointless anyway since no page seems to contain a playlist
* The VALID_URL regex is replaced by `(?P<id>[^/?#&]+)` as you said
* The test now assumes a mp4 to be returned instead of a m3u8.

File: youtube_dl/extractor/npo.py [hunk 7]
```diff
@@ -406,6 +406,27 @@ def _real_extract(self, url):
...
+            'title': 'Ademhaling: De hele dag haal je adem. Maar wat gebeurt er dan eigenlijk in je lichaam?',
+            'ext': 'mp4',
+            'description': 'md5:abfa0ff690adb73fd0297fd033aaa631'
+        }
+    }
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        webpage = self._download_webpage(url, playlist_id)
+        video_id = re.findall(r'VideoPlayer.*data-mid="([^"]+)"', webpage)[0]
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: Use `self._search_regex` instead.

**Comment by rrooij** [comment]
Done. I also changed the variable names since the id doesn't resemble a playlist id anymore.

File: youtube_dl/extractor/npo.py [hunk 8]
```diff
@@ -406,6 +406,27 @@ def _real_extract(self, url):
...
+            'title': 'Ademhaling: De hele dag haal je adem. Maar wat gebeurt er dan eigenlijk in je lichaam?',
+            'ext': 'mp4',
+            'description': 'md5:abfa0ff690adb73fd0297fd033aaa631'
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        npo_video_id = self._search_regex(r'data-mid="([^"]+)"', webpage, video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: 3rd argument should be the name of field you search for `'video id'`.

File: youtube_dl/extractor/npo.py [hunk 9]
```diff
@@ -406,6 +406,27 @@ def _real_extract(self, url):
...
+        'info_dict': {
+            'id': 'WO_NTR_429477',
+            'title': 'Ademhaling: De hele dag haal je adem. Maar wat gebeurt er dan eigenlijk in je lichaam?',
+            'ext': 'mp4',
+            'description': 'md5:abfa0ff690adb73fd0297fd033aaa631'
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: This is usually called `display_id` and included in info dict as well.

**Comment by rrooij** [comment]
I changed the variable to `display_id`. However, I didn't include it in the info_dict because the test fails. It expects the video_id instead (`WO_NTR_...`).

**Comment by dstftw** [comment]
I mean it should be included as `display_id` field.

**Comment by rrooij** [comment]
Did I do it correctly?

As far as I can see, the difference between the `_type` being `url_transparant` and `url` is that the former lets you override the info_dict of the other extractor? Is this correct? Since it does the display_id correctly now, instead of when I manually added the display_id to the dictionary result from `url_result`.

Thanks for your help by the way!

File: youtube_dl/extractor/npo.py [hunk 10]
```diff
@@ -406,6 +406,33 @@ def _real_extract(self, url):
...
+
+    _TEST = {
+        'url': 'http://www.schooltv.nl/video/ademhaling-de-hele-dag-haal-je-adem-maar-wat-gebeurt-er-dan-eigenlijk-in-je-lichaam/',
+        'info_dict': {
+            'id': 'WO_NTR_429477',
+            'display_id': 'ademhaling-de-hele-dag-haal-je-adem-maar-wat-gebeurt-er-dan-eigenlijk-in-je-lichaam',
+            'title': 'Ademhaling: De hele dag haal je adem. Maar wat gebeurt er dan eigenlijk in je lichaam?',
+            'ext': 'mp4',
+            'description': 'md5:abfa0ff690adb73fd0297fd033aaa631'
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/npo.py]: m3u8 downloads with ffmpeg should be [skipped](https://github.com/rg3/youtube-dl/commit/60ad3eb9706861d4182ab44ee19d64350ca2e36e).

**Comment by rrooij** [comment]
Done. And I understand, downloading the video is redundant in test cases if the json contains all the necessary information.

**Comment by dstftw** [comment]
No. ffmpeg or rtmpdump may stall making the whole test build stall.

**Comment by dstftw** [comment]
Thanks.

**Pull request merged**